client/cloudevents/sourceclient.go

```diff
@@ -0,0 +1,315 @@
+package cloudevents
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/google/uuid"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	"open-cluster-management.io/api/client/cloudevents/payload"
+	"open-cluster-management.io/api/client/cloudevents/protocol"
+	"open-cluster-management.io/api/client/cloudevents/types"
+)
+
+type CloudEventSourceClient[T ResourceObject] struct {
+	contextGetter    protocol.CloudEventsContextGetter
+	subClient        cloudevents.Client
+	pubClient        cloudevents.Client
+	lister           Lister[T]
+	codecs           map[types.GroupVersionResource]Codec[T]
+	statusHashGetter StatusHashGetter[T]
+	sourceID         string
+}
+
+func NewCloudEventSourceClient[T ResourceObject](
+	contextGetter protocol.CloudEventsContextGetter,
+	subClient, pubClient cloudevents.Client,
+	lister Lister[T],
+	statusHashGetter StatusHashGetter[T],
+	sourceID string,
+) *CloudEventSourceClient[T] {
+	return &CloudEventSourceClient[T]{
+		contextGetter:    contextGetter,
+		subClient:        subClient,
+		pubClient:        pubClient,
+		lister:           lister,
+		codecs:           make(map[types.GroupVersionResource]Codec[T]),
+		statusHashGetter: statusHashGetter,
+		sourceID:         sourceID,
+	}
+}
+
+func (c *CloudEventSourceClient[T]) WithCodec(gvr types.GroupVersionResource, codec Codec[T]) {
+	c.codecs[gvr] = codec
+}
+
+// Resync the resources status of the current source by sending a status resync request to agents
+func (c *CloudEventSourceClient[T]) Resync(ctx context.Context, eventGVR types.GroupVersionResource) error {
+	eventType := types.CloudEventType{
+		GroupVersionResource: eventGVR,
+		SubResource:          types.SubResourceStatus,
+		Action:               types.ResyncRequestAction,
+	}
+
+	evt := cloudevents.NewEvent()
```

Comment:
i think we should have a wrapper of NewEvent with generated uuid and time, and type/source as input.

Reply:
add a event builder for this